File: SOC/src/cc/game/soc/core/Card.java
package cc.game.soc.core;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.PrintWriter;

import cc.lib.utils.Reflector;

public class Card extends Reflector<Card> implements Comparable <Card> {
	
	static {
		addAllFields(Card.class);
	}
	
	private CardType type;
	private int typeOrdinal;
	private CardStatus status;

	@Override
	public void serialize(PrintWriter out) throws IOException {
		out.println(type+","+status+","+type.dereferenceOrdinal(typeOrdinal).name());
	}

	@Override
	protected void deserialize(BufferedReader in) throws Exception {
	    String [] parts = in.readLine().split(",");
		type = CardType.valueOf(parts[0]);
		status = CardStatus.valueOf(parts[1]);
		switch (type) {
			case Commodity:
				typeOrdinal = CommodityType.valueOf(parts[2]).ordinal();
				break;
			case Development:
				typeOrdinal = DevelopmentCardType.valueOf(parts[2]).ordinal();
				break;
			case Event:
				typeOrdinal = EventCardType.valueOf(parts[2]).ordinal();
				break;
			case Progress:
				typeOrdinal = ProgressCardType.valueOf(parts[2]).ordinal();
				break;
			case Resource:
				typeOrdinal = ResourceType.valueOf(parts[2]).ordinal();
				break;
			case SpecialVictory:
				typeOrdinal = SpecialVictoryType.valueOf(parts[2]).ordinal();
				break;
			default:
				throw new cc.lib.utils.GException("Unhandled case");
			
		}
	}

    @Override
    protected boolean isImmutable() {
        return true;
    }

    /**
	 * Needed for Reflector
	 */
	public Card() {
		this(null, 0, null);
	}
	
	/**
	 * 
	 * @param type
	 */
	public Card(ICardType<?> type) {
		this(type.getCardType(), type.ordinal(), CardStatus.USABLE);
	}
	
	/**
	 * 
	 * @param type
	 * @param status
	 */
	public Card(ICardType<?> type, CardStatus status) {
		this(type.getCardType(), type.ordinal(), status);
	}
	
	/**
	 * 
	 * @param type
	 * @param ordinal
	 * @param status
	 */
	public Card(CardType type, int ordinal, CardStatus status) {
		this.type = type;
		this.status = status;
		this.typeOrdinal = ordinal;
	}
	
	@Override
    public String toString() {
    	return type.dereferenceOrdinal(typeOrdinal).name() + " " + status;
    }

	/**
	 * 
	 * @return
	 */
    public final String getName(StringResource sr) {
    	return type.dereferenceOrdinal(typeOrdinal).getName(sr);
    }
    
    /**
     * 
     * @return
     */
	public final boolean isUsable() {
		return status == CardStatus.USABLE;
	}
	
	/**
	 * 
	 */
	public final void setUsable() {
		status = CardStatus.USABLE;
	}
	
	/**
	 * 
	 * @return
	 */
	public final boolean isUsed() {
		return status == CardStatus.USED;
	}
	
	/**
	 * 
	 */
	public final void setUsed() {
		status = CardStatus.USED;
	}
	
	/**
	 * 
	 */
	public final void setUnusable() {
		status = CardStatus.UNUSABLE;
	}
	
	/**
	 * @return Returns the type.
	 */
	public final CardType getCardType() {
		return type;
	}
	
	/**
	 * 
	 * @return
	 */
	public final CardStatus getCardStatus() {
		return this.status;
	}
	
	/**
	 * An ordinal into the type of card this card represents (development or progress)
	 * @return
	 */
	public final int getTypeOrdinal() {
		return typeOrdinal;
	}
	
	/**
	 * 
	 * @return
	 */
	public final String getHelpText(Rules rules, StringResource sr) {
		return type.dereferenceOrdinal(typeOrdinal).getHelpText(rules, sr);
	}
	
	/**
	 * 
	 * @return
	 */
	public Object getData() {
		return type.dereferenceOrdinal(typeOrdinal).getData();
	}
	
	/**
	 * 
	 * @param card
	 * @return
	 */
	public boolean equals(ICardType<?> card) {
		return (card.getCardType() == getCardType() && card.ordinal() == getTypeOrdinal());
	}
	
	@Override
	public boolean equals(Object obj) {
		if (obj == this)
			return true;
		if (!(obj instanceof Card))
			return false;
		Card card = (Card)obj;
		return type == card.type && typeOrdinal == card.typeOrdinal && status == card.status;
	}
	
	@Override
	public int compareTo(Card o) {
		if (type != o.type)
			return type.compareTo(o.type);
		if (typeOrdinal != o.typeOrdinal)
			return typeOrdinal - o.typeOrdinal;
		// we want usable cards to appear earliest in the list, then unusable, then used
		return status.compareTo(o.status);
	}


}

File: SOC/src/cc/game/soc/core/Card.kt
package cc.game.soc.core

import cc.lib.reflector.RBufferedReader
import cc.lib.reflector.RPrintWriter
import cc.lib.reflector.Reflector
import cc.lib.utils.GException
import java.io.IOException

open class Card(var cardType: CardType, var typeOrdinal: Int, var cardStatus: CardStatus)
	: Reflector<Card>(), Comparable<Card> {
	companion object {
		init {
			addAllFields(Card::class.java)
		}
	}

	constructor(type: ICardType<*>) : this(type.cardType, type.ordinal, type.defaultStatus())

	constructor(type: ICardType<*>, status: CardStatus) : this(type.cardType, type.ordinal, status)

	constructor() : this(CardType.Resource, 0, CardStatus.UNUSABLE)

	@Throws(IOException::class)
	public override fun serialize(out: RPrintWriter) {
		out.println(cardType.toString() + "," + cardStatus + "," + cardType.dereferenceOrdinal(typeOrdinal).name)
	}

	@Throws(Exception::class)
	override fun deserialize(`in`: RBufferedReader) {
		val parts = `in`.readLine().split(",".toRegex()).toTypedArray()
		cardType = CardType.valueOf(parts[0])
		cardStatus = CardStatus.valueOf(parts[1])
		typeOrdinal = when (cardType) {
			CardType.Commodity -> CommodityType.valueOf(parts[2]).ordinal
			CardType.Development -> DevelopmentCardType.valueOf(parts[2]).ordinal
			CardType.Event -> EventCardType.valueOf(parts[2]).ordinal
			CardType.Progress -> ProgressCardType.valueOf(parts[2]).ordinal
			CardType.Resource -> ResourceType.valueOf(parts[2]).ordinal
			CardType.SpecialVictory -> SpecialVictoryType.valueOf(parts[2]).ordinal
			else -> throw GException("Unhandled case")
		}
	}

	override fun isImmutable(): Boolean {
		return true
	}

	override fun toString(): String {
		return cardType.dereferenceOrdinal(typeOrdinal).getNameId() + " " + cardStatus
	}

	/**
	 *
	 * @return
	 */
	val name: String
		get() = cardType.dereferenceOrdinal(typeOrdinal).getNameId()

	/**
	 *
	 * @return
	 */
	val isUsable: Boolean
		get() = cardStatus === CardStatus.USABLE

	/**
	 *
	 */
	fun setUsable() {
		cardStatus = CardStatus.USABLE
	}

	/**
	 *
	 * @return
	 */
	val isUsed: Boolean
		get() = cardStatus === CardStatus.USED

	/**
	 *
	 */
	fun setUsed() {
		cardStatus = CardStatus.USED
	}

	/**
	 *
	 */
	fun setUnusable() {
		cardStatus = CardStatus.UNUSABLE
	}

	/**
	 *
	 * @return
	 */
	fun getHelpText(rules: Rules): String? {
		return cardType.dereferenceOrdinal(typeOrdinal).getHelpText(rules)
	}

	/**
	 *
	 * @return
	 */
	val data: Any?
		get() = cardType.dereferenceOrdinal(typeOrdinal).getData()

	override fun equals(obj: Any?): Boolean {
		if (obj === this) return true
		if (obj is ICardType<*>) {
			return obj.cardType == cardType && obj.ordinal == typeOrdinal
		}
		if (obj is Card)
			return cardType == obj.cardType && typeOrdinal == obj.typeOrdinal && cardStatus === obj.cardStatus
		return false
	}

	override fun compareTo(o: Card): Int {
		if (cardType !== o.cardType) return cardType.compareTo(o.cardType)
		return if (typeOrdinal != o.typeOrdinal) typeOrdinal - o.typeOrdinal else cardStatus.compareTo(o.cardStatus)
		// we want usable cards to appear earliest in the list, then unusable, then used
	}
	/**
	 *
	 * @param cardType
	 * @param typeOrdinal
	 * @param cardStatus
	 */
}